Content.Server/ShipEvent/Systems/ShipEventFactionSystem.cs

```diff
@@ -0,0 +1,148 @@
+            Teams![spawnerEntity].AddMember(shipEventRole);
+        }
+    }
+
+    private void CreateTeam(EntityUid spawnerEntity, bool silent = false)
+    {
+        (EntityUid shipGrid, string shipName) = GetShipData(spawnerEntity);
+        Teams![spawnerEntity] = new PlayerFaction(GenerateTeamName(), "/Textures/Theta/ShipEvent/ShipFactionIcon.rsi");
+        if(!silent)
+        {
+            Announce(Loc.GetString(
+                "shipevent-team-add",
+                ("teamname", Teams![spawnerEntity].Name),
+                ("shipname", shipName)));
+        }
+    }
+
+    private void RemoveTeam(EntityUid spawnerEntity, bool silent = false, string removeReason = "")
+    {
+        (EntityUid shipGrid, string shipName) = GetShipData(spawnerEntity);
+
+        if (!silent)
+        {
+            string _removeReason = Loc.GetString("shipevent-remove-default");
+            if (removeReason != "") { _removeReason = removeReason; }
+            string message = Loc.GetString(
+                "shipevent-team-remove",
+                ("teamname", Teams![spawnerEntity].Name),
+                ("shipname", shipName),
+                ("removereason", _removeReason));
+            Announce(message);
+        }
+
+        Teams!.Remove(spawnerEntity);
+        if (shipGrid != null)
+        {
+            entMan.DeleteEntity((EntityUid)shipGrid);
+            entMan.DeleteEntity(spawnerEntity);
+        }
+    }
+
+    private void CheckTeams()
+    {
+        foreach (EntityUid spawnerEntity in Teams!.Keys)
+        {
+            PlayerFaction faction = Teams![spawnerEntity];
+            if (faction.GetLivingMembers().Count == 0 && faction.Members.Any()) {
+                RemoveTeam(
+                spawnerEntity,
+                false,
+                Loc.GetString("shipevent-remove-dead"));
+            }
+        }
+    }
+
+    private void Announce(string message)
+    {
+        chatSystem.DispatchGlobalAnnouncement(message, Loc.GetString("shipevent-announcement-title"));
+    }
+
+    private string GenerateTeamName()
+    {
+        var taken = GetTakenNames();
+        while (true)
+        {
+            LastTeamNumber += 1;
+            string name = "Team №" + LastTeamNumber.ToString();
+            if (!taken.Contains(name)) { return name; }
+        }
+    }
+
+    private IEnumerable<string> GetTakenNames()
+    {
+        List<string> taken = new();
+        foreach (EntityUid spawnerEntity in Teams!.Keys) { taken.Add(Teams![spawnerEntity].Name); }
+        return taken;
+    }
+
+    private (EntityUid, string) GetShipData(EntityUid spawnerEntity)
+    {
+        TransformComponent transform = entMan.GetComponent<TransformComponent>(spawnerEntity);
+        EntityUid? shipGrid = transform.GridUid;
+        if (shipGrid == null) { return (EntityUid.Invalid, ""); }
+        string shipName = entMan.GetEntityQuery<MetaDataComponent>().GetComponent((EntityUid)shipGrid).EntityName;
```

Comment:
```suggestion
        string shipName = entMan.GetEntityQuery<MetaDataComponent>().GetComponent(shipGrid.Value).EntityName;
```
Там же не надо тайпкастить

Reply:
надо, он систем.нуллабл.ентитиуид а не просто ентитиуид

Reply:
пусть там и ретурн выше, он всё равно этого не понимает видимо

Reply:
странно